wallet/src/de/schildbach/wallet/ui/more/masternode_keys/MasternodeKeysViewModel.kt

```diff
@@ -40,25 +53,60 @@ import javax.inject.Inject
 class MasternodeKeysViewModel @Inject constructor(
     private val walletData: WalletDataProvider,
     private val clipboardManager: ClipboardManager,
-    private val securityFunctions: SecurityFunctions,
+    private val securityFunctions: SecurityFunctions
 ) : ViewModel() {
+    companion object {
+        private val log = LoggerFactory.getLogger(MasternodeKeysViewModel::class.java)
+    }
 
-    private val authenticationGroup: AuthenticationGroupExtension = walletData.wallet!!.addOrGetExistingExtension(AuthenticationGroupExtension(walletData.wallet)) as AuthenticationGroupExtension
+    private val authenticationGroup: AuthenticationGroupExtension = walletData.wallet!!.addOrGetExistingExtension(
+        AuthenticationGroupExtension(walletData.wallet)
+    ) as AuthenticationGroupExtension
     private val masternodeKeyChainTypes = EnumSet.of(
         AuthenticationKeyChain.KeyChainType.MASTERNODE_OWNER,
         AuthenticationKeyChain.KeyChainType.MASTERNODE_VOTING,
         AuthenticationKeyChain.KeyChainType.MASTERNODE_OPERATOR,
-        AuthenticationKeyChain.KeyChainType.MASTERNODE_PLATFORM_OPERATOR,
+        AuthenticationKeyChain.KeyChainType.MASTERNODE_PLATFORM_OPERATOR
     )
 
     val keyChainMap = hashMapOf<MasternodeKeyType, MasternodeKeyTypeInfo>()
 
     private val masternodeKeyChainInfoMap = hashMapOf<MasternodeKeyType, MasternodeKeyChainInfo>()
+    private val _newKeysUsed = MutableLiveData(false)
+    val newKeysFound: LiveData<Boolean>
+        get() = _newKeysUsed
 
     init {
         if (authenticationGroup.hasKeyChains()) {
             initKeyChainInfo()
         }
+
+        walletData.observeAuthenticationKeyUsage()
+            .onEach { usage -> refreshKeyChains(usage) }
+            .launchIn(viewModelScope)
+    }
+
+    private fun refreshKeyChains(usage: List<AuthenticationKeyUsage>) {
+        walletData.wallet?.let { wallet ->
+            val ownerPath = DerivationPathFactory.get(wallet.params).masternodeOwnerDerivationPath()
+            val rootPath = ownerPath.subList(0, ownerPath.size - 1)
+
+            // determine the number of new keys that were used
+            val masternodeKeysUsed = usage.count { keyUsage ->
+                if (keyUsage.key is IDeterministicKey) {
+                    val key = keyUsage.key as IDeterministicKey
+                    key.path.containsAll(rootPath)
+                } else {
+                    false
+                }
+            }
+            if (masternodeKeysUsed > 0) {
+                initKeyChainInfo()
+                _newKeysUsed.value = true
+            } else {
+                _newKeysUsed.value = false
+            }
```

Comment:
Here we are using liveData to tell the UI to update based on any changes in masternode key usage found during blockchain sync.

Reply:
Since the `refreshKeyChains` is executed in `viewModelScope`, it will run on UI thread. This is probably fine if the method executes fast.
If not, it's better to `launchIn` another scope, for example, a custom working scope that this viewModel will hold like this:

```
private val workerJob = SupervisorJob()
val viewModelWorkerScope = CoroutineScope(Dispatchers.IO + workerJob) 
```